keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1091 @@
+start. It checks that the hardware is still available and healthy, then
+configures it for the pod as specified in the allocation type.
+
+### Publishing node capacity
+
+A numeric model defines a capacity type. The type has its own API group, kind
+and potentially different versions. It must support encoding and decoding JSON
+with an apimachinery scheme. The kubelet queries the registered DRA driver
+kubelet plugins for capacity information through a new gRPC call. Drivers
+respond by returning the JSON encoding of this capacity type. A streaming gRPC
+API allows drivers to update the capacity information.
+
+For example, a simple counter could be described as:
+```
+const GroupName = "counter.dra.config.k8s.io"
+const KindName = "Capacity"
+const Version = "v1beta1"
+
+type Capacity struct {
+	metav1.ObjectMeta
+	metav1.TypeMeta
+
+	Count int64
+}
+```
+
+The kubelet then combines all of this information in a new NodeResourceCapacity
+object which contains the JSON representation of a capacity type in a
+`runtime.RawExtension`. The kubelet itself never has to interpret that data.
+This enables running a newer control plane and DRA driver while keeping kubelet
+at an older version.
+
+### Using numeric parameters as claim parameters
+
+The numeric model also defines a type for claim parameters. For the example
+above, that could be:
+```
+type Parameters struct {
+	metav1.ObjectMeta
+	metav1.TypeMeta
+
+	Required int64
+	Selector metav1.LabelSelector
+}
+```
+
+The `Parameters.Required` value gets compared against `Capacity.Count`. The
+`Selector` is matched with the labels defined in `Capacity.ObjectMeta`.
+
+When defining the CRD for the claim parameters, the driver developer replicates
+those fields from the numeric model which are relevant for the driver and adds
+custom validation. For a complex model, this can be a true subset of the fields
+in the parameter type. The validation can set bounds that are specific to the
+driver.
+
+```
+const GroupName = "dra.e2e.example.com"
+const KindName = "ClaimParameters"
+const Version = "v1alpha1"
+
+type ClaimParameters struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	metav1.TypeMeta   `json:",inline"`
+
+	Spec ClaimParameterSpec `json:"spec"`
+}
+
+type ClaimParameterSpec struct {
+	// +kubebuilder:default=1
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=16
+	NumberOfFrobnicators int64 `json:"numberOfFrobnicators,omitempty"`
+
+	// Additional configuration parameters, in this case
+	// some environment variables for the container in which
+	// the Frobnicators are going to be used.
+	Env map[string]string `json:"env,omitempty"`
+}
+```
+
+The following template is used to generate a
+`counter.dra.config.k8s.io/Parameters` description from a
+`dra.e2e.example.com/ClaimParameters` object:
+
+```
+kind: Parameters
+apiVersion: counter.dra.config.k8s.io/v1alpha1
+Required: {{object.spec.numberOfFrobnicators}}
+```
```

Comment:
@johnbelamaric: does this make sense?

I know that in your example you had instead something like this:
```
kind: Parameters
apiVersion: counter.dra.config.k8s.io/v1alpha1
Required: "object.spec.numberOfFrobnicators"
```

But I don't know how to implement that: this cannot be decoded into a `counter.Parameters` because the `Required` field is an int64. The CEL expression has to be evaluated and replaced *before* decoding.

We could decode into `unstructured.Unstructured`, but then how would the code determine which of the `any` values is a CEL expression string and what are literal strings?

We also can't make it a rule that only "leaf" values can be expressions. Suppose the driver author had wanted to populate the `Selector`. Then the template becomes:
```
kind: Parameters
apiVersion: counter.dra.config.k8s.io/v1alpha1
Required: {{object.spec.numberOfFrobnicators}}
Selector: {{object.spec.frobnicatorLabels}}
```

Here `object.spec.frobnicatorLabels` must be a valid value for the `metav1.LabelSelector` struct.

Reply:
So, in my [proposal](https://github.com/kubernetes/enhancements/pull/4384#issuecomment-1912886646), there are request object and their are request expression objects. They are separate data types. Think of it as a "strongly typed" version of a template:

```go
// Request objects
type ResourceRequest struct {
	Name string

	// exactly one should be populated
	Capability *bool
	Counter *ResourceCounterRequest
	Quantity *ResourceQuantityRequest
}

type ResourceCounterRequest struct {
	Request int64 `json:"request"`
}

type ResourceQuantityRequest struct {
	Request resource.Quantity `json:"request"`
}

// Request expression objects
type ResourceRequestExpression struct {
	Name string

	// exactly one should be populated
	Capability *bool
	Counter *ResourceCounterRequestExpression
	Quantity *ResourceQuantityRequestExpression
}

type ResourceCounterRequestExpression struct {
	Request string `json:"request"`
}

type ResourceQuantityRequestExpression struct {
	Request string `json:"request"`
}
```

So, no templating is needed because the structure of the types is built in.

An aside: I didn't bother with a CapabilityRequestExpression struct because it's just a boolean. But in fact I can imagine wanting to examine the CR and decide whether a given capability is required or not, so we might want one of those.

See also https://github.com/kubernetes/kubernetes/pull/123033#issuecomment-1917856708

Reply:
So with your approach, a vendor can inject content from their CRD into pre-defined "slots" of the request. What if those slots are for fields that are more than just simple string or integer values?

In this example, the entire `Selector` comes from the CRD:

```
kind: Parameters
apiVersion: counter.dra.config.k8s.io/v1alpha1
required: {{object.spec.numberOfFrobnicators}}
selector: {{object.spec.frobnicatorLabels}}
```

Now suppose that having a full `metav1.LabelSelector` is too complex for the user-facing API. Instead, the CRD contains individual fields that need to be converted to a `metav1.LabelSelector`. That's a better API because all available labels are documented directly in the CRD.

With a template, that's easy:

```
kind: Parameters
apiVersion: counter.dra.config.k8s.io/v1alpha1
required: {{object.spec.numberOfFrobnicators}}
selector:
  matchLabels:
    driver.example.com/frobnicatorColor: {{object.spec.frobnicatorColor}}
    driver.example.com/frobnicatorSmell: {{object.spec.frobnicatorSmell}}
```

With your approach, the CEL expression would need to construct a `metav1.LabelSelector`. Is that really possible?

I'm still using `metav1.LabelSelector` as example. The actual label selector will probably be even more complex.

Reply:
One more thought about
>  Think of it as a "strongly typed" version of a template

and

> Each expression must return the appropriate type, based upon the underlying capacity struct.

We cannot validate that because the input object has unknown types for its fields.

The net result is that there is more validation of a request expression object (some of the structure must be right) compared to a template string, but developers can still make mistakes which are only caught during evaluation against a specific CR.

Reply:
> The net result is that there is more validation of a request expression object (some of the structure must be right) compared to a template string, but developers can still make mistakes which are only caught during evaluation against a specific CR.

Yeah, agreed. We can evaluate that the expression is syntactically correct, and CEL will do impilicity type conversions. But you absolutely could still have runtime failures against specific CRs, if the author of the CEL expression makes a mistake.

Reply:
> With your approach, the CEL expression would need to construct a metav1.LabelSelector. Is that really possible?

Yes, I have done something similar in CEL to generate a `map[string]string`, it looked something like this:

```go
// MapExpr is used for various fields to calculate map entries. Only one of
// Key and KeyExpr may be specified; similarly only on of Value and ValueExpr
// may be specified.
type MapExpr struct {
	Key       *string `json:"key,omitempty"`
	Value     *string `json:"value,omitempty"`
	KeyExpr   *string `json:"keyExpr,omitempty"`
	ValueExpr *string `json:"valueExpr,omitempty"`
}
```

So, in yaml you could write:

```yaml
mapEntries:
- key: foo
  value: bar
- key: foobar
  labelExpr: "object.foobar"
- keyExpr: " 'foo-' + object.bar "
  labelExpr: "object.foo[object.bar]"
```

The result would be a map with three key/value pairs. This was a user-facing API so having both a literal and expression version of each of key and value was convenient. But it's semantically equivalent to only have expressions.

Patrick - this probably doesn't address your concern you raised in our live discussion about if the source has unknown numbers of entries in an array (ie, array -> array type processing). Will I find that example in the mode switch KEP? In general, there is still a lot of work to figure out how to express everything.

Reply:
> this probably doesn't address your concern you raised in our live discussion about if the source has unknown numbers of entries in an array (ie, array -> array type processing).

Indeed, that is the problematic case. Even worse, it might be recursive.

> Will I find that example in the mode switch KEP?

We can use the [label selector](https://github.com/kubernetes/enhancements/blob/dff3dfb8b5c1fa39afba13716f6e0cb6de8a1206/keps/sig-node/4383-dra-semantic-model-mode-switching/README.md#label-selector) from that KEP as target type and the example from https://github.com/NVIDIA/k8s-dra-driver/blob/main/demo/specs/selectors/parameters.yaml#L8 as input CR.

We need to transform this:

```
apiVersion: gpu.resource.nvidia.com/v1alpha1
kind: GpuClaimParameters
metadata:
  namespace: kubecon-demo
  name: "inference-gpu"
spec:
  selector:
    andExpression:
    - memory:
        value: "16G"
        operator: LessThanOrEqualTo
    - cudaComputeCapability:
        value: "7.5"
        operator: GreaterThanOrEqualTo
```

into this:

```
apiVersion: modeswitch.dra.config.k8s.io/v1alpha1
kind: Parameters
selector:
  and:
    - label: memory
      value: "16G"
      op: LEQ
      type: Quantity
    - label: cudaComputeCapability
      value: "7.5"
      type: Semver
```

This is just one example. We cannot assume that the input has exactly this structure. It might also be:
https://github.com/NVIDIA/k8s-dra-driver/blob/b6c7aae2b87d857668f417689462da752090406f/demo/specs/selectors/parameters.yaml#L18-L27

cc @klueska

Reply:
This last comment sets off alarms for me.  

a) Encoding an expression-grammar in YAML is kind of disgusting.
b) Encoding a vendor-specific expression-grammar in YAML is even worse.
c) Transforming one vendor-specific expression grammar into another expression grammar is even worse yet.

This feels a bit like "we made a wrong turn 3 km back".

Reply:
> a) Encoding an expression-grammar in YAML is kind of disgusting.

I was asking about a type for the expression (e.g. `metav1.LabelSelector`) with the corresponding more complex YAML vs string which then gets parsed back when I started [the thread about selectors](https://github.com/kubernetes/enhancements/pull/4384#discussion_r1455189444).

@thockin : do we have an example in the Kubernetes API where a string is used for a complex expression? I've only found `metav1.ListOptions.FieldFilter` and `LabelFilter`, but those are relatively simple key/value pairs.

What became clear is that we need the capability. How to expose it in the API is then the next step. The right place to discuss this is in https://github.com/kubernetes/enhancements/pull/4464 - that is the KEP which defines the selector, not this one here.

> b) Encoding a vendor-specific expression-grammar in YAML is even worse.

This is something for @klueska to comment on....

> c) Transforming one vendor-specific expression grammar into another expression grammar is even worse yet.
> 
> This feels a bit like "we made a wrong turn 3 km back".

Yes. Before even reading your comments I also came to the same conclusion. A session at the Gym with earphones forgotten at home can be very productive :smile: 

Where I took the wrong turn is with the decision to let the scheduler read vendor claim parameter objects and then have it somehow extract relevant information out of that. The motivation was to enable DRA drivers which only consist of a kubelet plugin (similar to a device plugin). That might be laudable goal (weak excuse, I know), but it leads to all this complexity around permissions, dynamically watching unknown resources, and last but least the whole question around how to extract that information (duck typing, templating, expression objects, etc.).

Let's take a step back. Here's my idea:
- The parameter types of each model become objects in the apiserver, each in their own API group (independent versioning!). Alternatively, we could also add `ClaimParameters` as a container for multiple parameter types to `resource.k8s.io`, but separate API groups looks cleaner and simpler to me. I need to think about whether `NodeResourceCapacity` should also live in those new API groups - this has implications for kubelet.
- The vendor needs to write a controller which translates each CR into the in-tree parameter objects. Those objects link back to the vendor CR with UID, group, kind, name, and revision.
- The scheduler watches those in-tree parameter objects. When it encounters a claim which references a vendor CR, it can find the corresponding in-tree parameter objects *and* can check that they are up-to-date by comparing the revision.

This can be combined with strongly-typed in-tree APIs.

@klueska: writing such a CRD controller wouldn't be a problem, would it?

Reply:
This sounds reasonable for a first step, for sure. It simplifies things a lot. I think it's possible to do as we have been dicussing but not necessary now.

Reply:
> The parameter types of each model become objects in the apiserver, each in their own API group (independent versioning!). Alternatively, we could also add ClaimParameters as a container for multiple parameter types to resource.k8s.io, but separate API groups looks cleaner and simpler to me. I need to think about whether NodeResourceCapacity should also live in those new API groups - this has implications for kubelet.

I do still think separate kinds for model parameters is overkill. A single kind that contains an array of union structs seems sufficient. I don't think independent versioning of the models is helpful, I think it just creates complexity.

If the concern is for custom schedulers that use custom models, we can always add a custom union type and put that complexity in that use case, rather than it appearing everywhere for everyone.

Reply:
I agree - I came to the same conclusion after thinking through it a bit more.

Reply:
I wasn't precise. I agree that we don't need separate kinds.

I disagree with independent versioning. We do need that because at some point we want to to have a resource.k8s.io v1beta1 API which includes a new "foobar" v1alpha1 semantic model. This should be reflected in the Go packages which define the semantic models. Using different Go packages also makes it clearer what belongs together and what doesn't. We can make the final decision during the implementation review.

Reply:
I'm going to ignore the discussion of semantics as types vs builtins - it's a distraction unless we have a clear(ish) model in mind.

> do we have an example in the Kubernetes API where a string is used for a complex expression?

All the places we adopt CEL fit this bill.  See `ValidatingAdmissionPolicySpec`

>> b) Encoding a vendor-specific expression-grammar in YAML is even worse.
>
> This is something for @klueska to comment on....

Not pointing any fingers - we're all figuring this out together, and frankly, we're now willing to reconsider constraints that we may have imposed in the past, so our options are better.

e.g.  Assume that we can agree that "contraints" and "requests" are seperable.  The "selector" as you called it could be a CEL expression (using CEL builtins and k8s extensions) which is run over the published attributes of each resource, producing a boolean result (or even a score!).

Scheduler looks at a pending pod, finds candidate nodes based on hard-criteria like node selector, CPU, memory, existence of requested resource types, then runs that CEL against the potential resources to see which nodes+devices satisfy.

THEN, finally, we can look at the actual resources requested (the vendor CR).  We still need to map fields in the CR to fields in the Cpacaity, but if we're excluding constraints, there should be far fewer of them, so maybe we could map them by fieldpath?

Downside: We can syntax-check CEL but we can't know if it referenced fields are typo'ed or if it is trying to multiply strings.

Upside: We can approximate the evaluation cost of CEL and we can extend it with functions, approximating a DSL.

> The vendor needs to write a controller which translates each CR into the in-tree parameter objects.

Help me understand how the autoscaler uses this?  

A pod is pending which requests an "acme.com/frobnicator" and 5 "FUs" (frobber units!).  The autoscaler can't act on that until the controller updates the user's CR to translate that into a ks8-neutral form, right?

Reply:
I still don't know how to think about parallel modes which require a significant configuration operation which affects all potential consumers.

E.g. Hypothetical device which can do power-of-two sizes.  I can slice a card with 8 units into {1*8U} or {2*4U} or {1*4U, 2*2U} or {1*4U, 1*2U, 2*1U}.  For the very first pod, we have to make a decision, which isn't so bad (we'll be wrong, but OK).  But how do I know what the current mode is when I want to schedule a second pod?

When the kubelet publishes capacity, does it just dump all the parallel options with all the params?  Can I end up with 8 cards, each with 16 virtual functions, each with 4 possible slice-ups, each with 10 attributes (for constraints) and 2 requestable quantities, yielding 5000+ entries in the schema?

Reply:
> > This is something for @klueska to comment on....
> 
> Not pointing any fingers - we're all figuring this out together, and frankly, we're now willing to reconsider constraints that we may have imposed in the past, so our options are better.

I was just punting on this because he is the expert on real world use cases, not me. He also has the feedback from customers about how they want to select hardware, which might address the "is this a good user interface" aspect.

> > The vendor needs to write a controller which translates each CR into the in-tree parameter objects.
> 
> Help me understand how the autoscaler uses this?
> 
> A pod is pending which requests an "acme.com/frobnicator" and 5 "FUs" (frobber units!). The autoscaler can't act on that until the controller updates the user's CR to translate that into a ks8-neutral form, right?

Correct. But the translation is done by a control plane component of the DRA driver, not by the individual DRA kubelet plugin, so this translation will happen pretty much at the point where the user creates the vendor parameter CR. The translation *only* has that parameter CR as input, not the pod or claim (both not known yet) and not anything from a node (that's what the output gets compared again later).

> But how do I know what the current mode is when I want to schedule a second pod?

The scheduler records the chosen mode in the allocation result (as permanent record) and in it's internal state. Then when it looks at another pod, it knows the mode of an instance.

> When the kubelet publishes capacity, does it just dump all the parallel options with all the params? Can I end up with 8 cards, each with 16 virtual functions, each with 4 possible slice-ups, each with 10 attributes (for constraints) and 2 requestable quantities, yielding 5000+ entries in the schema?

Potentially yes. But if the number of combinations explodes like that, then we clearly need a different way of modeling valid combinations. @klueska came up with a vendor CRD for capacity and claim parameters that get matched against that capacity information. This tells me that it is doable for relevant use cases. How well it then fits other areas I don't know.

I tried to capture my (probably incomplete) understanding in https://github.com/kubernetes/enhancements/pull/4464. That might be a better place for a more focused discussion around the specific semantic.